src/components/checkbox/checkbox.jsx

```diff
@@ -0,0 +1,132 @@
+import React, {
+  forwardRef,
+  useEffect,
+  useImperativeHandle,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'react';
+import PropTypes from 'prop-types';
+import FormControl from '../form-control/form-control.jsx';
+import useMounted from '../../hooks/use-mounted.js';
+import Icon from '../icon/icon.jsx';
+import styles from './checkbox.css';
+import cautionSvg from '../../svgs/caution.svg';
+
+const Checkbox = forwardRef(function Checkbox(props, forwardedRef) {
+  const fallbackRef = useRef();
+  const ref = forwardedRef || fallbackRef;
+  const inputRef = useRef();
+  const mounted = useMounted();
+  const [validationMessage, setValidationMessage] = useState(props.validationMessage);
+
+  useEffect(() => {
+    if (!mounted.current) return;
+
+    inputRef.current.checked = props.checked;
+  }, [props.checked]);
+
+  useEffect(() => {
+    setValidationMessage(props.validationMessage);
+  }, [props.validationMessage]);
+
+  useLayoutEffect(() => {
+    inputRef.current.setCustomValidity(validationMessage);
+  }, [validationMessage]);
```

Comment:
All of the validation logic (including the `onBlur` and `onChange` events) are pretty similar to what was copied from the `Date` component.

I started with the `useValidation` hook, but since `value` is `on`/`off` instead of falsey or truthy, the internals didn't work.

I spent some time on an extracted version of _these_ events as a hook and planned on backfilling it to date, but it seemed like having two `validation` hooks felt bad / weird. The commit is here (ed6f7b0ba2ae309d5831f8dd2b915aa2faa472eb) but I reverted it.

Reply:
Yeah, I wanted to see how the new validation spec holds up in a few implementations before committing to redoing the validation hook.